src/groupeddataframe/grouping.jl

```diff
@@ -181,6 +197,242 @@ end
+    show(io, NamedTuple(k))
+end
+
+Base.parent(key::GroupKey) = getfield(key, :parent)
+Base.length(key::GroupKey{N}) where N = length(N)
+Base.names(::GroupKey{N}) where N = N
+Base.keys(key::GroupKey) = names(key)
+Base.values(key::GroupKey{N,T}) where {N,T} = _groupvalues(parent(key), getfield(key, :idx))::T
+
+Base.iterate(key::GroupKey) = iterate(key, 1)
+Base.iterate(key::GroupKey, i::Integer) = i <= length(key) ? (key[i], i + 1) : nothing
+
+Base.getindex(key::GroupKey, i::Integer) = _groupvalue(parent(key), getfield(key, :idx), i)
+
+function Base.getindex(key::GroupKey{N}, n::Symbol) where N
+    i = findfirst(==(n), N)
+    i === nothing && throw(KeyError(n))
+    return key[i]
+end
+function Base.getproperty(key::GroupKey, p::Symbol)
+    try
+        return key[p]
+    catch e
+        throw(ErrorException("$(typeof(key)) has no property $p"))
+    end
+end
+
+Base.:(==)(k1::GroupKey, k2::GroupKey) = parent(k1) === parent(k2) && getfield(k1, :idx) == getfield(k2, :idx)
+
+Base.NamedTuple(key::GroupKey{N}) where {N} = NamedTuple{N}(values(key))
+Base.Tuple(key::GroupKey) = values(key)
+
+
+"""
+    keys(gd::GroupedDataFrame)
+
+Get the set of keys for each group in `GroupedDataFrame` object `gd`.
+The keys are [`GroupKey`](@ref) objects which behave like `NamedTuple`s holding
+the values of the grouping keys and can be used to index into `gd` efficiently.
+
+### Returns
+
+Vector of [`GroupKey`](@ref)s.
+
+### Examples
+
+```jldoctest
+julia> df = DataFrame(a = repeat([:foo, :bar, :baz], outer=[4]),
+                      b = repeat([2, 1], outer=[6]),
+                      c = 1:12);
+
+julia> gd = groupby(df, [:a, :b])
+GroupedDataFrame with 6 groups based on keys: a, b
+First Group (2 rows): a = :foo, b = 2
+│ Row │ a      │ b     │ c     │
+│     │ Symbol │ Int64 │ Int64 │
+├─────┼────────┼───────┼───────┤
+│ 1   │ foo    │ 2     │ 1     │
+│ 2   │ foo    │ 2     │ 7     │
+⋮
+Last Group (2 rows): a = :baz, b = 1
+│ Row │ a      │ b     │ c     │
+│     │ Symbol │ Int64 │ Int64 │
+├─────┼────────┼───────┼───────┤
+│ 1   │ baz    │ 1     │ 6     │
+│ 2   │ baz    │ 1     │ 12    │
+
+julia> keys(gd)
+6-element Array{DataFrames.GroupKey{(:a, :b),Tuple{Symbol,Int64}},1}:
+ GroupKey(a = :foo, b = 2)
+ GroupKey(a = :bar, b = 1)
+ GroupKey(a = :baz, b = 2)
+ GroupKey(a = :foo, b = 1)
+ GroupKey(a = :bar, b = 2)
+ GroupKey(a = :baz, b = 1)
+
+julia> k = keys(gd)[1];
+
+julia> keys(k)
+(:a, :b)
+
+julia> values(k)
+(:foo, 2)
+
+julia> NamedTuple(k)
+(a = :foo, b = 2)
+
+julia> k.a
+:foo
+
+julia> k[:a]
+:foo
+
+julia> k[1]
+:foo
+
+julia> gd[k]
+2×3 SubDataFrame
+│ Row │ a      │ b     │ c     │
+│     │ Symbol │ Int64 │ Int64 │
+├─────┼────────┼───────┼───────┤
+│ 1   │ foo    │ 2     │ 1     │
+│ 2   │ foo    │ 2     │ 7     │
+```
+"""
+function Base.keys(gd::GroupedDataFrame)
+    N = Tuple(_names(gd)[i] for i in gd.cols)
+    V = Tuple{_grouptypes(gd)...}
+    return [GroupKey{N, V}(gd, i) for i in 1:length(gd)]
```

Comment:
OK - maybe here is a benefit of making `GroupKey` parametric on `N` and `V` - right?

Reply:
Yes I guess in the case of a loop over `keys(gd)`, the compiler may be able to realize that `N` and `V` don't change across `GroupKey` objects and infer the type of the groups. But that needs to be checked.

Reply:
I think you would get the benefit but you have to use a barrier function to get it. This is a speculation though.

@jlumpe - have you tested this? (probably using `@inferred` inside a barier function could be used in a test)

Reply:
Yes and for the function barrier to have an effect, `keys(gd)` would have to return an object parameterized on the column types. Probably not worth it though.

Reply:
@nalimilan The return type of `keys(gd)` is already parameterized by the column types?

Reply:
Yes it is now, but if you create `keys` and do not use barier function then you get no benefit of it, because the compiler will not know these types at compile time anyway. So the question is (as posted above) - is it likely that users will write the code with barier function + even if they do then this barier function will need to be recompiled each time a new `groupby` is performed (due to type stability issues). So my concern is (and I assume also @nalimilan shares it) is this beneficial to have it in a typical usage scenario (for sure if you have millions of groups and use a barier function and within it you heavily use the return value of `keys` then there is a benefit, but the question is if this is a typical use case).

Reply:
Good question. I mostly did it because it seemed more correct and wasn't much extra work. I'm not sure if there will be much performance benefit as is, but I think that once a typed version of `DataFrame` is implemented this will need to be done anyways.

Now that I think of it, it might be beneficial to parameterize `GroupedDataFrame` instead. That should be much easier than for `DataFrame` and shouldn't be backwards incompatible. This information would then be available to `GroupKey` via the type parameter for its parent, which I just added.

Reply:
As long as `DataFrame` and `GroupedDataFrame` are not type-stable, probably better not parameterize the new types either unless you can see performance improvements in benchmarks. Parameters can actually make things slower if they are not inferred. These can always be added later if we make `DataFrame` type-stable at some point.

Reply:
I agree. Also my intuition for the future is that type-stability in `DataFrame` will be optional (I am not sure how it will be achieved but this is a discussion for post 1.0 release).

Reply:
Makes sense, I removed all parameters but for the parent `GroupedDataFrame`.